Add softPwm.h bindings using updated bindgen

lennartS opened the pull request:
These new bindings are still based on the same version, but now they
include the functions from softPwm.h as well. Some syntactic sugar
was added for software PWM pins just like for the other pins.

In this first implementation the range of software PWM is fixed to 100,
so PWM values can can range from 0 to 100. This value is based on the
official documentation.

These software PWM pins are useful since the raspberry pi only has one
hardware PWM controller.

In addition, the bindings were generated using bindgen version 0.19.1.

Signed-off-by: Lennart Sauerbeck <devel@lennart.sauerbeck.org>

lennartS committed:
Add softPwm.h bindings using updated bindgen

These new bindings are still based on the same version, but now they
include the functions from softPwm.h as well. Some syntactic sugar
was added for software PWM pins just like for the other pins.

In this first implementation the range of software PWM is fixed to 100,
so PWM values can can range from 0 to 100. This value is based on the
official documentation.

These software PWM pins are useful since the raspberry pi only has one
hardware PWM controller.

In addition, the bindings were generated using bindgen version 0.19.1.

Signed-off-by: Lennart Sauerbeck <devel@lennart.sauerbeck.org>

lennartS commented:
One slightly ugly thing about this pull request is the combination of updated bindgen version and softwarePwm.h introduction. My rationale for this was that the bindings.rs is autogenerated and as such a beautiful diff is not that important.

If you'd like I can split this commit into two, the first one with only the updated bindgen version and the second one with the introduction of softwarePwm.h

Razican commented:
It looks fine by me. Is it formatted with `rustfmt`?

@Ogeon, what do you think?

lennartS commented:
Neither `bindings.rs` nor `lib.rs` are formatted using `rustfmt`. `bindings.rs` is generated anyway and `lib.rs` has some (pre-existing) comments that are too long.

That being said, splitting up the comments and running `rustfmt` should be easy, just tell me if you'd want me to run it.

Ogeon commented:
I have been avoiding `rustfmt` until I can turn off the visual indenting (it messes with the git diffs), so no need, in my opinion, as long as it looks good. I'll read through it now.

Ogeon commented:
Yep, this looks really good! I just wonder what `softPwmStop` does, exactly, since it's used when converting to softpwm. Oh, and a bit of documentation would be nice, as well. It doesn't have to be anything fancy. [This (possibly outdated) page](https://projects.drogon.net/raspberry-pi/wiringpi/software-pwm-library/) seem to have some things that may be good to bring forwards.

lennartS commented:
Very good point about the documentation. I moved that behind playing around with it and then just forgot about it.

In addition to the documentation I also added an example and cleaned up the calls to `softPwmStop`. They are only needed in the conversion calls from `SoftPwmPin`, not in `into_soft_pwm()`.

I also implented `Drop` for `SoftPwmPin` which calls `softPwmStop`. Though I ran into some trouble with the `&mut self`, can you please have a look at the `Drop` implementation to see if that's how it's supposed to be done?

lennartS commented:
@Ogeon, not sure if you're subscribed to this, so notifying you directly.

I added some documentation as well as an example. Could you take another look at it?

Ogeon commented on src/lib.rs:
```diff
@@ -224,6 +229,92 @@ pub mod pin {
+        /// constant high) resulting in a frequenzy of 100 Hz.
+        ///
+        /// **Important**: In order to use software PWM pins *wiringPi*
+        /// has to be setup in GPIO mode via `setup_gpio()`.
```
It may be a good idea to move some of this info to the struct description, to put it at the top of the page, to make it more visible.

lennartS replied:
I agree and have done so in the next commit.

Ogeon commented on src/lib.rs:
```diff
@@ -224,6 +229,92 @@ pub mod pin {
+        }
+    }
+
+    impl<P> Drop for SoftPwmPin<P> {
```
I'm not sure how good idea it is to implement `Drop`, since there are no guarantees that this is the only reference to this pin, even if it's discouraged. I'm also thinking that it may also be a bit surprising if they just "stop working" when it drops out of scope.

Ogeon replied:
@Razican, what do you think about this? You have most surely used this library more recently than I have.

lennartS replied:
I liked it at first but your point of view makes a lot of sense. Even if it is the only reference it is out of line with the other pins, which do not implement `Drop`.

So even if it would be a good idea (which would be a separate discussion) `Drop` should be added for all pins in a separate commit.

I removed it in the next commit.

Ogeon replied:
Alright. Well, `Drop` could have made more sense if this library was smarter, but currently there's no tracking of anything, apart from what each pin object has (which is essentially just its type). That separate discussion about whether all of them should have it would hopefully start out as a redesign/extension that makes it all smarter and safer. :smile:

Ogeon reviewed (commented):
Oh, sorry about that. I got caught up in other things and forgot that I hadn't responded. I have added a couple of comments, but it's otherwise great!

Ogeon reviewed (changes requested):
Nice! I think this is more or less ready to go, but it seems like you forgot to add a stop method, now that `Drop` is gone. I think that's all! (I'm also trying out the review tool. Not too bad!)

lennartS commented on src/lib.rs:
```diff
@@ -269,6 +269,16 @@ pub mod pin {
 
+        /// Stops the software handling of this pin.
+        ///
+        /// _Note_: In order to control this pin via software PWM again
```
This note is a bit redundant since `pwm_stop` consumes `self` anyway. However, it states the intent more clearly and as a new-ish rust user I like this redundancy.

lennartS commented:
Yes, you're right about a stop method. I added it. I also added an inline comment to the first commit about the note I put in the documentation (which does not seem to show up in the conversation here). Let me know if you'd rather have the note removed.

In a second commit I then changed the conversion methods to use the stop method to avoid code duplication (even if it is just three lines).

Ogeon commented:
Great! Thank you!

> Let me know if you'd rather have the note removed.

No, it's a useful hint. It doesn't hurt to be extra clear. :smile:

Let me know when you feel ready (I see that you squashed it) and I'll merge.

lennartS commented:
Yeah, I went ahead with the squash after I saw you approved the request.

I think it's ready, you can go ahead with the merge.

Let me use this opportunity to thank both of you, @Ogeon and @Razican, for the work on this library, it made working with the Pi much easier :) And thanks for the patience and hints during this pull request, @Ogeon.

Ogeon commented:
I'm glad to hear that, and thank you for your contribution! Feel free to do it again :smile:
